pkg/virt-launcher/virtwrap/util/libvirt_helper.go

```diff
@@ -251,71 +254,120 @@ func StartLibvirt(stopChan chan struct{}) {
 			if err != nil {
-				log.Log.Reason(err).Error("failed to start virtlogd")
-				panic(err)
+				errMsg := fmt.Sprintf("failed to open logfile in path: \"%s\". Error: %s", logfile, err.Error())
+				log.Log.Reason(err).Error(errMsg)
+				return
 			}
+			defer util.CloseIOAndCheckErr(file, nil)
 
-			go func() {
-				logfile := fmt.Sprintf("/var/log/libvirt/qemu/%s.log", domainName)
-
-				// It can take a few seconds to the log file to be created
-				for {
-					_, err = os.Stat(logfile)
-					if !os.IsNotExist(err) {
-						break
-					}
-					time.Sleep(time.Second)
-				}
-				// #nosec No risk for path injection. logfile has a static basedir
-				file, err := os.Open(logfile)
-				if err != nil {
-					log.Log.Reason(err).Error("failed to catch virtlogd logs")
-					return
-				}
-				defer util.CloseIOAndCheckErr(file, nil)
+			scanner := bufio.NewScanner(file)
+			scanner.Buffer(make([]byte, 1024), 512*1024)
+			for scanner.Scan() {
+				log.LogQemuLogLine(log.Log, scanner.Text())
+			}
 
-				scanner := bufio.NewScanner(file)
-				scanner.Buffer(make([]byte, 1024), 512*1024)
-				for scanner.Scan() {
-					log.LogQemuLogLine(log.Log, scanner.Text())
-				}
+			if err := scanner.Err(); err != nil {
+				log.Log.Reason(err).Error("failed to read virtlogd logs")
+			}
+		}()
+
+		go func() {
+			defer close(exitChan)
+			_ = cmd.Wait()
+		}()
+
+		select {
+		case <-stopChan:
+			_ = cmd.Process.Kill()
+			return
+		case <-exitChan:
+			log.Log.Errorf("virtlogd exited, restarting")
+		}
 
-				if err := scanner.Err(); err != nil {
-					log.Log.Reason(err).Error("failed to read virtlogd logs")
-				}
-			}()
+		// this sleep is to avoid consumming all resources in the
+		// event of a virtlogd crash loop.
+		time.Sleep(time.Second)
+	}
+}
 
-			go func() {
-				defer close(exitChan)
-				cmd.Wait()
-			}()
+func startQEMUSeaBiosLogging(stopChan chan struct{}) {
+	const QEMUSeaBiosDebugPipeMode uint32 = 0666
+	const logLinePrefix = "[SeaBios]:"
+
+	err := syscall.Mkfifo(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed creating a pipe for sea bios debug logs: %s", logLinePrefix, err.Error()))
+		return
+	}
+
+	// Chmod is needed since umask is 0018. Therefore Mkfifo does not actually create a pipe with proper permissions.
+	err = syscall.Chmod(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed executing chmod on pipe for sea bios debug logs: %s", logLinePrefix, err.Error()))
+		return
+	}
+
+	QEMUPipe, err := os.OpenFile(QEMUSeaBiosDebugPipe, os.O_RDONLY, 0666)
+
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed to open %s with an error: %s", logLinePrefix, QEMUSeaBiosDebugPipe, err.Error()))
+	}
+	defer QEMUPipe.Close()
+
+	var line []byte
+	reader := bufio.NewReader(QEMUPipe)
+
+	for {
+		for line, _, err = reader.ReadLine(); err == nil; line, _, err = reader.ReadLine() {
```

Comment:
I got confused here :)
do you read this twice on purpose ? `line, _, err = reader.ReadLine()`

Reply:
This wasn't my intention. I thought what happens is (similar to C++/Java for loops):
* Before first iteration: `line, _, err` are being initialized with `reader.ReadLine()`
* After each iteration break if `err != nil`
* Before starting a new iteration, update `line, _, err` by calling `reader.ReadLine()`.

So as I see it `line` isn't being read twice, but once for initialization and another time for each iteration.
Or am I missing something here?

Reply:
@iholder-redhat I don't think your for loop is doing what you intended. In general the golang loop is like this :
```
for [ Initial Statement ] ; [ Condition ] ; [ Post Statement ] {
}
```
And this is not aligned with your assumptions.  I suggest to just do the Redline inside the loop and then check the error, if EOF then break , if another error return, if not an error do the process. 

BTW, are you familiar with golang playground?   If you are not sure about behavior you can easily write a very small piece of code and test it there directly ... For example you could tested the loop behavior

Reply:
@itamarh I think your loop will work correctly (your description is correct), but it's indeed a little confusing :-) 
Usually people will just do the readline inside the loop body, and add explicit tests for error (return) and EOF (break), which might be clearer.

Reply:
I'm just curious why did you choose to use a ReadLine? Wouldn't it be simpler to use a scanner? 
Maybe something like this will work?
```
var line string
 scanner := bufio.NewScanner(file)                                           
    if scanner.Scan() {                                                         
        line = scanner.Text()                                                 
    }                                                                           
    if err := scanner.Err(); err != nil {                                       
        return nil, err                                                         
    }
```

Reply:
@vladikr Don't you need to have `for scanner.Scan()  { `  ?

Reply:
oh yes, of course :)
sorry

Reply:
@ezrasilvera I see what you mean. It is a bit confusing :) It's just that I usually don't prefer loops which lack a concrete condition as it's more error prone as it can lead to infinite loops, i.e.
```
for {
    line, err := ...
    if err != nil {
        break
    }
}
```
Maybe the line read can occur inside the loop (and one time outside of it for initialization) and we'll check the error as a condition for the loop? something like:
```
line, err := ...
for err == nil {
    line, err := ...
}

if err != nil {
    return ...
}
```
IMHO although possible, I'm not sure it's better since, at least in my eyes, it's less obvious to figure out what the for loop is actually doing in the above approach (since the loop body is only error checking).
As I see it what I did is pretty standard - moved the initialization line (`line, err := ...`), the condition (`err == nil`) and the post statement (`line, err := ...`) into the loop body.

Just my opinion :) what do you guys think?

Reply:
@vladikr I can change the implementation to use Scanner. I've read a bit about the differences between Scanner and Reader but found only minor differences while the internet says that for the majority of cases they're interchangeable. As I'm still new to Go can you explain why in your opinion Scanner is better for this case? :) thanks!

Reply:
it's just ReadLine is more low level and as you saw requires more handling. `Scanner` is more general-purpose and incorporates some of this logic.
you can see that the golang page also doesn't recommend ReadLine: https://golang.org/pkg/bufio/#Reader.ReadLine

In any case, I don't have a strong opinion, the look just caught my eye and it took me some time to understand. I'm sure people will start questioning this next time they'll read it :)
I don't want us to stall on this. If it works, either way, I'm +1 :)

Reply:
@vladikr 
I don't have a strong opinion on this either, just wanted to educate myself on the reason for using Scanner instead. If even golang page recommends Scanner I guess it's the right choice. Also I obviously agree that code shouldn't be confusing.

I'll change this to using Scanner and re-upload :) thanks for the answer.